### PR TITLE
index.Finder.Nearest: support lt and gt

### PR DESCRIPTION
**Pull request by mattnibs**
Change the name of closet to nearest, make it a public method and
support > and < operations. Get rid of ClosetGTE/LTE and have users just
call nearest with the correct operator.

Part of #3183

### CODE REVIEW

File: index/finder.go [hunk 1]
```diff
@@ -45,72 +55,69 @@ func NewFinder(ctx context.Context, zctx *zed.Context, engine storage.Engine, ur
...
-			if op == eql {
-				rec = nil
-			}
-			if op == gte {
+			if op == GTE || op == GT {
 				return prev, nil
 			}
-			return rec, nil
+			if op == EQL {
+				return nil, nil
```
Comment (nwt) [inline index/finder.go:116]: Should this case remain consistent with the corresponding case in lookupAsc? (They were previously identical but now diverge.)

File: index/finder.go [hunk 2]
```diff
@@ -231,7 +214,13 @@ func compareFn(kvs []KeyValue) expr.KeyCompareFn {
 	}
 }
 
-func (f *Finder) closest(kvs []KeyValue, op operator) (*zed.Value, error) {
+// Nearest finds the zed.Value in the index that is nearest to kvs according to
+// the provided operator.
```
Comment (nwt) [inline index/finder.go:218]:
```suggestion
// operator.
```

File: index/finder.go [hunk 3]
```diff
@@ -270,3 +259,21 @@ func (f *Finder) ParseKeys(inputs ...string) ([]KeyValue, error) {
...
 	return kvs, nil
 }
+
+func (o Operator) hasEqual() bool {
+	switch o {
+	case EQL, GTE, LTE:
+		return true
+	default:
+		return false
+	}
```
Comment (nwt) [inline index/finder.go:269]:
```suggestion
	}
	return false
```

File: index/finder.go [hunk 4]
```diff
@@ -270,3 +259,21 @@ func (f *Finder) ParseKeys(inputs ...string) ([]KeyValue, error) {
...
+}
+
+func (o Operator) valid() bool {
+	switch o {
+	case EQL, GT, GTE, LT, LTE:
+		return true
+	default:
+		return false
+	}
```
Comment (nwt) [inline index/finder.go:278]:
```suggestion
	}
	return false
```

File: index/index_test.go [hunk 5]
```diff
@@ -85,59 +85,51 @@ func TestCompare(t *testing.T) {
...
-			}
-
+			rec, err := finder.Nearest(op, kvs...)
 			require.NoError(t, err)
-			if expected == -1 {
-				assert.Nil(t, rec)
-			} else {
-				require.NotNil(t, rec)
-				v, err := rec.AccessInt("ts")
+			var v int64 = -1
```
Comment (nwt) [inline index/index_test.go:103]:
```suggestion
			v := int64(-1)
```

**Review by nwt — APPROVED** [review]

File: index/finder.go (hunk 1 above)
Comment (mattnibs) [reply index/finder.go:115]: They are the same thing but I'll change the one above to this.

File: index/finder.go [hunk 6]
```diff
@@ -45,72 +55,68 @@ func NewFinder(ctx context.Context, zctx *zed.Context, engine storage.Engine, ur
...
-			if cmp == 0 {
+			if cmp == 0 && op.hasEqual() {
 				return rec, nil
 			}
-			if op == eql {
+			if op == LTE || op == LT {
+				return prev, nil
+			}
+			if op == EQL {
 				rec = nil
```
Comment (mattnibs) [inline index/finder.go:87]: return nil, nil

**Pull request merged**